ttnn/cpp/ttnn/operations/ccl/all_gather_matmul/device/multi_core/all_gather_matmul_op_multi_core.cpp

```diff
@@ -0,0 +1,167 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+///
+#include <algorithm>
+
+#include "tt_metal/common/core_coord.h"
+#include "eth_l1_address_map.h"
+#include "impl/buffers/buffer.hpp"
+#include "ttnn/tensor/tensor_impl.hpp"
+#include "ttnn/operations/ccl/all_gather/device/all_gather_op.hpp"
+#include "ttnn/operations/ccl/shared_with_host/hetergeneous_data_structs.hpp"
+#include "ttnn/operations/ccl/ccl_host_datastructures.hpp"
+#include "ttnn/operations/ccl/ccl_common.hpp"
+#include "ttnn/deprecated/tt_dnn/op_library/math.hpp"
+#include "ttnn/deprecated/tt_dnn/op_library/work_split.hpp"
+#include "tt_metal/common/constants.hpp"
+#include "tt_metal/detail/util.hpp"
+#include "tt_metal/host_api.hpp"
+#include <sstream>
+#include <type_traits>
+
+#include "ttnn/cpp/ttnn/operations/ccl/all_gather_matmul/device/all_gather_matmul_op.hpp"
+
+
+using namespace tt::constants;
+
+namespace ttnn {
+
+using namespace ccl;
+
+
+std::tuple<std::vector<CoreCoord>, std::vector<uint32_t>> setup_datacopy(
+    tt::tt_metal::Program& program,
+    const Tensor& input_tensor,
+    const Tensor& all_gather_output_tensor,
+    Tensor& datacopy_output_tensor,
+    const uint32_t dim,
+    const uint32_t num_links,
+    const uint32_t ring_size,
+    const uint32_t ring_index,
+    all_gather_op::Topology topology,
+
+    CoreCoord datacopy_core_coord
+) {
+
+    auto const& all_gather_config = AllGatherConfig(input_tensor, all_gather_output_tensor, dim, ring_size, num_links, topology);
+    const uint32_t num_transfers = 4; // ring_size - 1;
+
+    auto tensor_slicer = ttnn::ccl::InterleavedRingAllGatherTensorSlicer (
+        input_tensor,
+        all_gather_output_tensor,
+        dim,
+        ring_index
+    );
+
+
+    // Setup cores used for datacopy
+    std::vector<CoreCoord> all_datacopy_cores;
```

Comment:
we should move this down and assign it to the output of `corerange_to_cores` directly. It looks unnecessary how it's written write now.